site/js/modules/webui/server.js

```diff
@@ -0,0 +1,190 @@
+import * as Autograder from '../autograder/base.js'
+
+import * as Render from './render.js'
+import * as Routing from './routing.js'
+
+/* The priority of the field to show first. */
+/* Items later in the list have the highest priority. */
+const fieldPriority = [
+    "assignment-id",
+    "course-id",
+    "user-pass",
+    "user-email",
+];
+
+function init() {
+    Routing.addRoute(/^server$/, handlerServer, 'Server Actions', undefined);
+    Routing.addRoute(/^server\/call-api$/, handlerCallAPI, 'Call API', undefined);
+}
+
+function handlerServer(path, params, context, container) {
+    Routing.loadingStart(container);
+
+    let args = {
+        [Routing.PARAM_TARGET_ENDPOINT]: params[Routing.PARAM_TARGET_ENDPOINT],
+    };
+
+    let cards = [
+        Render.makeCardObject('server-action', 'Call API', Routing.formHashPath(Routing.PATH_SERVER_CALL_API, args)),
+    ];
+
+    container.innerHTML = `
+        ${Render.cards(cards)}
+    `;
+}
+
+function handlerCallAPI(path, params, context, container) {
+    Routing.loadingStart(container);
+
+    Autograder.Server.describe()
+        .then(function(result) {
+            const endpoints = result["endpoints"];
+            const selectedEndpoint = params[Routing.PARAM_TARGET_ENDPOINT] ?? undefined;
+
+            render(endpoints, selectedEndpoint, context, container);
+        })
+        .catch(function(message) {
+            console.error(message);
+            container.innerHTML = Render.autograderError(message);
+        })
+    ;
+}
+
+function render(endpoints, selectedEndpoint, context, container) {
+    let selector = renderSelector(endpoints, selectedEndpoint);
+    let endpointArea = renderEndpointArea(endpoints, selectedEndpoint, context);
+
+    let html = `
+        <div class="page-controls">${selector}</div>
+        <div class="endpoint-area">${endpointArea}</div>
+        <div class="results-area"></div>
+    `;
+
+    container.innerHTML = html;
+
+    container.querySelector(".page-controls select").addEventListener("change", function(event) {
+        let newParams = {
+            [Routing.PARAM_TARGET_ENDPOINT]: event.target.value,
+        };
+
+        let path = Routing.formHashPath(Routing.PATH_SERVER_CALL_API, newParams);
+        Routing.redirect(path);
+    });
+
+    let button = container.querySelector(".endpoint-area button");
+
+    if (button) {
+        button.addEventListener("click", function(event) {
+            callEndpoint(selectedEndpoint, endpoints[selectedEndpoint]["input"], context, container);
+        });
+    }
```

Comment:
You don't need to use it, but you could also use optional chaining instead of the conditional:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining